### ISSUE TITLE
CheckM error in binning genomes - python3

### ISSUE THREAD

**Issue by anwarMZ** [Ecogenomics/CheckM#250]
Hi I have installed CheckM v1.1.2 using anaconda python 3.6.7. I am trying to use 38 genomes using 32 threads in batch and somehow I keep getting an error for 9 processes and thus only 30 of 38 genomes are placed in bins

Command used
`checkm lineage_wf ./genomes/ -x fasta ./checkm_revision_muza/ -t 32 --ali --nt --pplacer_threads 64`

Error
`[2020-03-27 16:56:23] INFO: CheckM v1.1.2`
`[2020-03-27 16:56:23] INFO: checkm lineage_wf ./genomes/ -x fasta ./checkm_revision_muza/ -t 32 --ali --nt --pplacer_threads 64`
`[2020-03-27 16:56:23] INFO: [CheckM - tree] Placing bins in reference genome tree.`
`[2020-03-27 16:56:25] INFO: Identifying marker genes in 38 bins with 32 threads:`
`Process Process-16:sing 30 of 38 (78.95%) bins.`
`Traceback (most recent call last):`
  `File "/software/anaconda3/envs/wgs_py3/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap`
   `self.run()`
  `File "/software/anaconda3/envs/wgs_py3/lib/python3.6/multiprocessing/process.py", line 93, in run`
    `self._target(*self._args, **self._kwargs)`
 ` File "/software/anaconda3/envs/wgs_py3/lib/python3.6/site-packages/checkm/markerGeneFinder.py", line 115, in __processBin`
    `prodigal.run(binFile, bNucORFs)`
  `File "/software/anaconda3/envs/wgs_py3/lib/python3.6/site-packages/checkm/prodigal.py", line 104, in run`
    `prodigalParser = ProdigalGeneFeatureParser(gffFile)`
  `File "/software/anaconda3/envs/wgs_py3/lib/python3.6/site-packages/checkm/prodigal.py", line 191, in __init__`
    `self.__parseGFF(filename)`
  `File "/software/anaconda3/envs/wgs_py3/lib/python3.6/site-packages/checkm/prodigal.py", line 220, in __parseGFF`
    `start = int(lineSplit[3])`
`IndexError: list index out of range`

All files in the genome folder are `*.fasta` thus the `-x` flag. 
Versions of each of the system requirements are
`- HMMER = 3.1b2`
`- prodigal = 2.6.3`
`- pplacer = 1.1.alpha19`
`- numpy = 1.17.3`
`- pysam = 0.15.3`

- [ I also installed CheckM with pip3 and same error was produced. However, when same files were used in `CheckM v1.0.12` installed on another system with python 2 and this error wasn't thrown.] Could you please suggest?

Thanks, 
Zohaib

**Comment by dparks1134** [comment]
Hi Zohaib. My best guess is that one of the genomes isn't being parsed correctly by Prodigal.  Can you send the 38 genomes to me at donovan.parks [at] gmail.com so I can investigate?

**Comment by anwarMZ** [comment]
Hi Donovan, Thank you for prompt reply. I will have to ask the data owner first before sending since I didn't generate the data myself. Meanwhile, I was wondering if you can suggest what could be the parsing error and I can investigate along? Do you suggest running prodigal out of checkm on the genomes first? Also, same files work well when run with `CheckM v1.0.12`in Python2. A bit strange.

**Comment by dparks1134** [comment]
Hi. CheckM v1.1.2 is a port to Python 3. The estimated completeness and contamination is not expected to change compared to v1.0.12. The issue seems related to Prodigal based on where the code is breaking so it could be informative to run Prodigal v2.6.3 outside of CheckM on each genome. Any change you are using a newer version of Prodigal compared to when you ran CheckM v1.0.12? That said, Prodigal is stable so I wouldn't expect a new version to cause an issue.

**Comment by anwarMZ** [comment]
Hi again, I checked running prodigal out of `CheckM v1.1.2` and I can see that it works well for all genomes and I also saw discrepency in number of genes predicted in 8 genomes that were giving errors in CheckM. I believe thats because of merging the output of prodigal using multiple threads? I sense that because I was looking at output and all the genomes that had error had files ending with `genes.gff.4`, `genes.faa.4` & `genes.fna.4` unlike other bins. see screenshot attached. 

<img width="527" alt="Screenshot 2020-03-27 at 20 24 23" src="https://user-images.githubusercontent.com/25245737/77793111-d1aeeb00-7069-11ea-8f3d-c545dfbe9926.png">

Also first three lines of errors point to multiprocessing in `process.py`. 
`File "/software/anaconda3/envs/wgs_py3/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap`
`self.run()`
`File "/software/anaconda3/envs/wgs_py3/lib/python3.6/multiprocessing/process.py", line 93, in run`
`self._target(*self._args, **self._kwargs)`

Checked the Prodigal version in Python2 installation and it was same `Prodigal v2.6.3`

**Comment by dparks1134** [comment]
Hi. CheckM attemps to guess the correct translation table for genomes. The `.4` are results for translation table 4. I think these results just indicate CheckM crashed before testing table 11 and determine which translation table is most appropriate.

**Comment by anwarMZ** [comment]
On this basis could you suggest where to debug further? It seems like its not a Prodigal error then and instead parsing or "a checkpoint" crashes in CheckM. any suggestions?

**Comment by dparks1134** [comment]
I'm not sure what the issue is. Nobody else has report an issue with CheckM v1.1.2 so I'm assuming the issue is somehow specific to your system or set of genomes. As such, I'd begin debugging by running each genome through CheckM separately to establish which genome(s) are causing the issue. I'd then inspect this genome to see if anything is unusual and verify that the remaining genomes can run through CheckM as a single batch. This would at least narrow the issue down to a small subset (hopefully just one) genome and confirm the issue is with the input data and not CheckM per see. Of course, it could be a CheckM issue but my base assumption would be to assume the input data.

**Comment by anwarMZ** [comment]
I agree and I also think there is some discrepancy in genome files or intermediate files created. I will run it for each genome and see how it works out. Thank you for prompt helping. I will get back to you once I have some more information. I might need some help from you in debugging if the issue is narrow down to specific genomes. Cheers

**Comment by anwarMZ** [comment]
Hi again Donovan,
As discussed, I was looking at each genome individually over the weekend and figured out 8 genomes give same error. When those 8 genomes were taken out of pool, CheckM worked well for rest of 31 genomes. To remind you that with CheckM `python2.7 version` it works very well.
For investigating the error, I have added two FASTA files of publicly available genomes here. For better comparison, I have added one of the genomes that work well and one from error pool. 

`S_bockelmannii_VOUX01.1.fasta` gives error
`S_ficaria_QTUE01000001.fasta` works well.

I hope you can help with this error. 

Cheers, 
Zohaib
[Genomes_Checkm.zip](https://github.com/Ecogenomics/CheckM/files/4475368/Genomes_Checkm.zip)

**Comment by dparks1134** [comment]
Hi. Thanks for narrowing the issue down. I should be able to look at this tomorrow. Any chance all the problematic genomes had two periods "." in their filename?

**Comment by anwarMZ** [comment]
Hi Sorry for responding a bit late, that doesn't seem to be the case with two periods. 
5 "error producing" genomes belong to the set that have only one period in their name (a set of total 31 genomes). 2 "error producing" genomes belong to the set that have two periods in the name (a set of 7 genomes).

**Comment by dparks1134** [comment]
Hi,

The issue is that the erroneous file is using Windows-style end-of-line characters. This results in Prodigal producing an incorrectly formatted GFF file. 

Current output:
```
# Sequence Data: seqnum=26;seqlen=60087;seqhdr="VOUX01000026 Serratia marcescens strain S3 NODE_15_length_60087_cov_81.318464, whole genome shotgun sequence.
"
# Model Data: version=Prodigal.v2.6.3;run_type=Single;model="Ab initio";gc_cont=59.04;transl_table=4;uses_sd=1
```

Correct output:
```
# Sequence Data: seqnum=26;seqlen=60087;seqhdr="VOUX01000026 Serratia marcescens strain S3 NODE_15_length_60087_cov_81.318464, whole genome shotgun sequence."
# Model Data: version=Prodigal.v2.6.3;run_type=Single;model="Ab initio";gc_cont=59.04;transl_table=4;uses_sd=1
```

If you run `fromdos` or some other program to convert the end-of-line characters to those expected by Linux the genomes should run as expected. It would be appreciated if you could post this issue on the Prodigal GitHub:
https://github.com/hyattpd/Prodigal

Cheers,
Donovan

**Comment by anwarMZ** [comment]
Hi, 

Thank you for the insight. I will try to run the erroneous genome files in `fromdos` and try running again. 
I will also post on Prodigal GitHub too if it still gives problem. But out of curiosity, can I ask what would be the difference in CheckM `Python 2.7v` that still works well with same `Prodigal 2.6.3v` with these erroneous files?

Any idea what changed in reading these genomes?
Zohaib

**Comment by dparks1134** [comment]
Good question. It is a bit of a mystery to me. I can confirm that the `erroneous` genome does run though CheckM v1.0.0 (Python 2) on my end. I can also confirm, that Prodigal is still producing the incorrectly formatted GFF files in this case. The CheckM estimates are the same in both v1.0.0 and v1.1.2 so at least the CheckM results are consistent. So, ultimately, not sure what it is about v1.1.2 and/or Python 3 that causes a hard fail, though a hard fail makes more sense given that the file format is not as expected.

**Comment by anwarMZ** [comment]
Just to add more information towards debugging. I ran `fromdos` on all the genome files and those 7 erroneous genome files showed a difference in size (expectedly). Then running `CheckM v1.1.2` on all genome files in `Python 3` as wanted. Seems like, so far it's going well and hasn't produced any error at same stage where previously it used to. Although the `pplacer` step is still running. 

My understanding is that somewhere in reading the Fasta file in `Python 2` and `Python 3` is different when it comes to end of line character. And the error is already picked on in early stage but fails around `Prodigal` due to GFF file format. I tried running `Prodigal` out of `CheckM` and it works fine even when the genomes were not converted using `fromdos`

PS. maybe I should post this as a separate query for you? seems like `pplacer` runs only with 1 thread even though used `--pplacer_threads 64` parameter
